QDMA/linux-kernel/apps/dma-utils/dmactl.c

```diff
@@ -176,11 +176,13 @@ static int xnl_connect(struct xnl_cb *cb, int vf, void (*log_err)(const char *))
 
 	if (vf) {
         	attr->nla_len = strlen(XNL_NAME_VF) + 1 + NLA_HDRLEN;
-        	strcpy((char *)(attr + 1), XNL_NAME_VF);
+            // Use memcpy to avoid a fortify source error
+        	memcpy((char *)(attr + 1), XNL_NAME_VF, strlen(XNL_NAME_VF) + 1);
```

Comment:
I don't think you should use memcpy here but strscpy, checking the return value, and on older kernels that don't have strscpy, strlcpy.

Reply:
Basically, fortify source was making this stop because it couldn't find out the length of the string due to all the pointer mangling.

For what this is worth, this is user land code.

Reply:
>For what this is worth, this is user land code.

my bad. Then strlcpy would do I guess. Does fortify not allow them ?

Reply:
So Debian (Ubunut 18.04) seems to build with the following flags:

```
cc -g -O2 -fdebug-prefix-map=/builds/dma_ip_drivers=. -fstack-protector-strong -Wformat -Werror=format-security  -I. -I../include -I../dma-utils  -c -std=c99 -o ../dma-utils/dmaxfer.o ../dma-utils/dmaxfer.c -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE
```

I don't think `strlcpy` is in the c library (not linux's but maybe BSD's).

I had tried `strncpy` at first, and it doesn't seem to help. The compiler still complains, and upon execution, you will get a hard crash in your program.

Reply:
My bad again, since you are copying a constant you'll want to use snprintf.

Reply:
and for strlcpy : #include <bsd/string.h>

Reply:
I don't think it really matters what you use,

```bash
cc -g -g -O2  -fstack-protector-strong -Wformat -Werror=format-security -I. -I../include -I../dma-utils  -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE -c -std=c99 -o ../dma-utils/dmactl.o ../dma-utils/dmactl.c -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE
../dma-utils/dmactl.c: In function ‘xnl_common_msg_send’:
../dma-utils/dmactl.c:185:65: warning: ‘%s’ directive output truncated writing 6 bytes into a region of size 0 [-Wformat-truncation=]
  185 |          snprintf((char *)(attr + 1), strlen(XNL_NAME_PF) + 1, "%s", XNL_NAME_PF);
      |                                                                 ^~
In file included from /usr/include/stdio.h:867,
                 from ../dma-utils/dmactl.c:13:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output 7 bytes into a destination of size 0
   67 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../dma-utils/dmactl.c:180:65: warning: ‘%s’ directive output truncated writing 6 bytes into a region of size 0 [-Wformat-truncation=]
  180 |          snprintf((char *)(attr + 1), strlen(XNL_NAME_VF) + 1, "%s", XNL_NAME_VF);
      |                                                                 ^~
In file included from /usr/include/stdio.h:867,
                 from ../dma-utils/dmactl.c:13:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output 7 bytes into a destination of size 0
   67 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc -g -g -O2  -fstack-protector-strong -Wformat -Werror=format-security -I. -I../include -I../dma-utils  -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE -c -std=c99 -o ../dma-utils/dmactl_reg.o ../dma-utils/dmactl_reg.c -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE
../dma-utils/dmactl_reg.c: In function ‘print_repeated_reg.constprop’:
../dma-utils/dmactl_reg.c:224:35: warning: ‘%-47s’ directive output may be truncated writing between 47 and 127 bytes into a region of size between 87 and 90 [-Wformat-truncation=]
  224 |   snprintf(reg_dump, 100, "[%#7x] %-47s %#-10x %u\n",
      |                                   ^~~~~
  225 |    addr, name, val, val);
      |          ~~~~                      
In file included from /usr/include/stdio.h:867,
                 from ../dma-utils/dmactl_reg.c:18:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output between 72 and 164 bytes into a destination of size 100
   67 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

When i wrote my own library for this, I was using https://www.infradead.org/~tgr/libnl/ to ensure that I obtained the correct pointers. That said, I did go through Xilinx's arithmetic, and it seemed correct on this stage.

Reply:
For reference, here is the patch to this branch to recreate things more easily

<details>
```patch
diff --git a/QDMA/linux-kernel/apps/dma-ctl/Makefile b/QDMA/linux-kernel/apps/dma-ctl/Makefile
index a535a52..14d69ff 100644
--- a/QDMA/linux-kernel/apps/dma-ctl/Makefile
+++ b/QDMA/linux-kernel/apps/dma-ctl/Makefile
@@ -2,8 +2,9 @@ SHELL = /bin/bash
 
 CFLAGS += -g
 #CFLAGS += -O2 -fno-inline -Wall -Wstrict-prototypes
+CFLAGS += -g -O2  -fstack-protector-strong -Wformat -Werror=format-security
 CFLAGS += -I. -I../include -I../dma-utils
-CFLAGS += $(EXTRA_FLAGS)
+CFLAGS += $(EXTRA_FLAGS) -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGE_FILE_SOURCE -D_AIO_AIX_SOURCE
 
 DMA-CTL = dma-ctl
 DMA-UTILS_OBJS := $(patsubst %.c,%.o,$(wildcard ../dma-utils/*.c))
diff --git a/QDMA/linux-kernel/apps/dma-utils/dmactl.c b/QDMA/linux-kernel/apps/dma-utils/dmactl.c
index 72b8b66..44d2dee 100644
--- a/QDMA/linux-kernel/apps/dma-utils/dmactl.c
+++ b/QDMA/linux-kernel/apps/dma-utils/dmactl.c
@@ -177,12 +177,12 @@ static int xnl_connect(struct xnl_cb *cb, int vf, void (*log_err)(const char *))
        if (vf) {
                attr->nla_len = strlen(XNL_NAME_VF) + 1 + NLA_HDRLEN;
             // Use memcpy to avoid a fortify source error
-               memcpy((char *)(attr + 1), XNL_NAME_VF, strlen(XNL_NAME_VF) + 1);
+               snprintf((char *)(attr + 1), strlen(XNL_NAME_VF) + 1, "%s", XNL_NAME_VF);
 
        } else {
                attr->nla_len = strlen(XNL_NAME_PF) + 1 + NLA_HDRLEN;
             // Use memcpy to avoid a fortify source error
-               memcpy((char *)(attr + 1), XNL_NAME_PF, strlen(XNL_NAME_PF) + 1);
+               snprintf((char *)(attr + 1), strlen(XNL_NAME_PF) + 1, "%s", XNL_NAME_PF);
        }
         hdr->n.nlmsg_len += NLMSG_ALIGN(attr->nla_len);
 

```
</details>

```
cd QDMA/linux-kernel/apps
make -C dma-ctl
```

should cause the same warnings.